src/network/protocol/messages/merkleblock.zig

```diff
@@ -0,0 +1,146 @@
+pub const MerkleBlockMessage = struct {
+    block_header: BlockHeader,
+    transaction_count: u32,
+    hash_count: CompactSizeUint,
+    hashes: []const [32]u8,
+    flag_bytes: CompactSizeUint,
+    flags: []const u8,
+
+    const Self = @This();
+
+    pub inline fn name() *const [12]u8 {
+        return protocol.CommandNames.MERKLEBLOCK ++ [_]u8{0} ** 6;
+    }
+
+    /// Returns the message checksum
+    pub fn checksum(self: *const Self) [4]u8 {
+        var digest: [32]u8 = undefined;
+        var hasher = Sha256.init(.{});
+        self.serializeToWriter(hasher.writer()) catch unreachable;
+        hasher.final(&digest);
+
+        Sha256.hash(&digest, &digest, .{});
+
+        return digest[0..4].*;
+    }
+
+    /// Free the allocated memory
+    pub fn deinit(self: *const Self, allocator: std.mem.Allocator) void {
+        allocator.free(self.flags);
+        allocator.free(self.hashes);
+    }
+
+    /// Serialize the message as bytes and write them to the Writer.
+    pub fn serializeToWriter(self: *const Self, w: anytype) !void {
+        try self.block_header.serializeToWriter(w);
+        try w.writeInt(u32, self.transaction_count, .little);
+        try self.hash_count.encodeToWriter(w);
+        for (self.hashes) |hash| {
+            try w.writeAll(&hash);
+        }
+        try self.flag_bytes.encodeToWriter(w);
+        try w.writeAll(self.flags);
+    }
+
+    /// Serialize a message as bytes and return them.
+    pub fn serialize(self: *const Self, allocator: std.mem.Allocator) ![]u8 {
+        var list = std.ArrayList(u8).init(allocator);
+        errdefer list.deinit();
+        try self.serializeToWriter(list.writer());
+        return list.toOwnedSlice();
+    }
+
+    /// Returns the hint of the serialized length of the message.
+    pub fn hintSerializedLen(self: *const Self) usize {
+        // 80 bytes for the block header, 4 bytes for the transaction count
+        const fixed_length = 84;
+        const hash_count_len: usize = self.hash_count.hint_encoded_len();
+        const compact_hashes_len = 32 * self.hashes.len;
+        const flags_byte_ken: usize = self.flag_bytes.hint_encoded_len();
+        const flags_len = self.flags.len;
+        const variable_length = hash_count_len + compact_hashes_len + flags_byte_ken + flags_len;
+        return fixed_length + variable_length;
+    }
+
+    /// Deserialize a `MerkleBlockMessage` from a Reader.
+    pub fn deserializeReader(allocator: std.mem.Allocator, r: anytype) !Self {
+        var mb: Self = undefined;
+        mb.block_header = try BlockHeader.deserializeReader(r);
+        mb.transaction_count = try r.readInt(u32, .little);
+        mb.hash_count = try CompactSizeUint.decodeReader(r);
+
+        const hash_count = mb.hash_count.value();
+        const hashes = try allocator.alloc([32]u8, hash_count);
+        errdefer allocator.free(hashes);
+
+        for (hashes) |*hash| {
+            try r.readNoEof(hash);
+        }
```

Comment:
you can do one single big read for the whole `hashes` slice

Reply:
I don't think it's possible since hashes is 2d slice

Reply:
correct

Reply:
yet, does it have to be a 2D slice?

Reply:
maybe we can let it like this for now and improve on this late when we see the real usage we are doing of it